README.md
[![nlm-chat](https://img.shields.io/badge/chat-http%3A%2F%2Fsignup.testiumjs.com%2F-F4D03F?logo=chat&logoColor=white)](http://signup.testiumjs.com/)
[![nlm-github](https://img.shields.io/badge/github-testiumjs%2Ftestium--example--app%2Fissues-F4D03F?logo=github&logoColor=white)](https://github.com/testiumjs/testium-example-app/issues)
![nlm-node](https://img.shields.io/badge/node-%3E%3D12-blue?logo=node.js&logoColor=white)
![nlm-version](https://img.shields.io/badge/version-2.4.0-blue?logo=version&logoColor=white)
# Testium Example App

This application is used to test testium itself.
Changes to this application should be done carefully.
If additional behavior is required, it might be better to add a new file or endpoint than to change an existing one.

* Start on default port (4003): `npm start`
* Start on any port (e.g. 8080): `PORT=8080 npm start`


## Endpoints

### `/echo`

Sends a JSON response with information about the incoming request:

```js
{
  ip, // remote address/ip of the sender
  method, // the http method, e.g. 'GET'
  url, // the request url path, e.g. '/echo?a=42'
  body, // the request body for POST etc.
  headers, // the request headers
}
```


### `/error`

Sends a minimal error page with status code 500.


### `/crash`

Resets the connection without sending a valid response.
Simulates the app crashing.


### `/blackhole`

Never sends a response and keeps the connection open indefinitely.
Useful for testing timeout behavior, making sure the test suite can recover.


### `/blackholed-resource.html`

Simulates a resource required by the page (a script tag) never fully loading,
preventing the page load from finishing.


### `/dynamic.html`

Simulates a page that delays displaying and hiding certain elements,
e.g. because it's waiting for additional data to load.

* `div.load_later`: Shows after 300ms
* `div.load_never`: Is and stays hidden
* `div.hide_later`: Gets hidden after 300ms
* `div.hide_never`: Is and stays visible


### `/index.html`

A relatively complex page:

* Multiple `div.message` elements to verify uniqueness constraints for selectors
* Exactly one `div.only` to have a unique element to select
* `a.link-to-other-page` linking to `/other-page.html`
* A form with various input elements
* A button that triggers log messages
* Different alerts, triggered by links


### `/index-diff.html`

Looks just like `/index.html`, just with slightly different content at the bottom.
Can be used to test visual diffing.


### `/other-page.html`

A mostly empty page that is linked to by `/index.html`.


### `/redirect-after.html`

Uses a `meta` tag to redirect to `/index.html` after a few ms.


### `/redirect-to-query.html`

Just like `/redirect-after.html` but additionally passes two query params:

* `a b`: `A B` - to test encoding of spaces
* `c`: `1,7` - to test url encoding of special characters


### `/windows.html`

For testing popup & iframe related features.

#### `/frame.html`

Displayed as `iframe#cool-frame` and contains `div.in-iframe-only`
and `iframe#nested-frame` (see `/nested-frame.html`).

#### `/nested-frame.html`

This page contains `div#nested-frame-div`.

#### `/popup.html`

Can be opened by clicking on `button#open-popup` and contains `div.popup-only`.

#### `/draggable.html`

Displays a draggable element for testing buttonDown/Up and movePointer...

#### `/pwa.html`

Displays an installable minimal PWA testing page (it passes lighthouse audit with a score of 100)


CHANGELOG.md
### v2.4.0 (2021-07-19)
<a id="v2.4.0"></a>
#### 🏡 Internal

* [#21](https://github.com/testiumjs/testium-example-app/pull/21) add pwa test page ([@aotarola](https://github.com/aotarola))


### v2.3.0 (2021-07-19)
<a id="v2.3.0"></a>
#### 🏡 Internal

* [#20](https://github.com/testiumjs/testium-example-app/pull/20) migrate to gh actions ([@aotarola](https://github.com/aotarola))
* [`9fe0684`](https://github.com/testiumjs/testium-example-app/commit/9fe06845b66086fcbb9f5ef2fd36ceb448d8f695) chore: downgrade nlm version


### 2.2.4 - 2021-04-01

* chore(deps): Bump y18n from 4.0.0 to 4.0.1 - **[@dependabot[bot]](https://github.com/apps/dependabot)** [#12](https://github.com/testiumjs/testium-example-app/pull/12)
  - [`2a57189`](https://github.com/testiumjs/testium-example-app/commit/2a571891749dace74ef4be2bed4801694320e27a) **chore:** Bump y18n from 4.0.0 to 4.0.1 - see: [y18](Bumps [y18n](https://github.com/yargs/y18), [y18](- [Release notes](https://github.com/yargs/y18), [y18](- [Changelog](https://github.com/yargs/y18), [y18](- [Commits](https://github.com/yargs/y18)


### 2.2.3 - 2021-02-17

* chore(deps): Bump ini from 1.3.5 to 1.3.8 - **[@dependabot[bot]](https://github.com/apps/dependabot)** [#11](https://github.com/testiumjs/testium-example-app/pull/11)
  - [`28fc44d`](https://github.com/testiumjs/testium-example-app/commit/28fc44d7c1f95254f4a7a8d8688603e0aa26eda3) **chore:** Bump ini from 1.3.5 to 1.3.8 - see: [v1](- [Commits](https://github.com/isaacs/ini/compare/v1)


### 2.2.2 - 2020-07-20

* Bump lodash from 4.17.15 to 4.17.19 - **[@dependabot[bot]](https://github.com/apps/dependabot)** [#10](https://github.com/testiumjs/testium-example-app/pull/10)
  - [`b7b54b8`](https://github.com/testiumjs/testium-example-app/commit/b7b54b8970a02ae1114bf98c6747cfe2cf7b1dae) **chore:** Bump lodash from 4.17.15 to 4.17.19 - see: [4](- [Commits](https://github.com/lodash/lodash/compare/4)
  - [`42853c1`](https://github.com/testiumjs/testium-example-app/commit/42853c16dd5a2d952740ae03edbd506caa036e09) **docs:** update CONTRIBUTING.md


### 2.2.1 - 2020-04-16

* fix: transparent pixel src - **[@aaarichter](https://github.com/aaarichter)** [#9](https://github.com/testiumjs/testium-example-app/pull/9)
  - [`fc2aee9`](https://github.com/testiumjs/testium-example-app/commit/fc2aee99c03063786dac658ee81305e6117d2d82) **fix:** transparent pixel src


### 2.2.0 - 2020-04-15

* feat: add image test page & drop node 6 support - **[@aaarichter](https://github.com/aaarichter)** [#8](https://github.com/testiumjs/testium-example-app/pull/8)
  - [`5dba6bc`](https://github.com/testiumjs/testium-example-app/commit/5dba6bc746c2e652e2d6b390bbf75bb223fdab53) **refactor:** drop node support for Node 6 and below
  - [`e255fff`](https://github.com/testiumjs/testium-example-app/commit/e255fff35d24ce02d62922522b2d5b2fe499055c) **feat:** add image test page


### 2.1.1

* fix: node-static module is broken; remove it - **[@dbushong](https://github.com/dbushong)** [#7](https://github.com/testiumjs/testium-example-app/pull/7)
  - [`dc3dfb2`](https://github.com/testiumjs/testium-example-app/commit/dc3dfb2f7b4454a1a0598bdb08964433ddacb1e5) **fix:** node-static module is broken; remove it


### 2.1.0

* feat: add a draggable element test page - **[@dbushong](https://github.com/dbushong)** [#6](https://github.com/testiumjs/testium-example-app/pull/6)
  - [`8252fd6`](https://github.com/testiumjs/testium-example-app/commit/8252fd6015ca2505719739359837d1a9b734d3d1) **feat:** add a draggable element test page


### 2.0.2

* chore: Ship public and server to npm - **[@jkrems](https://github.com/jkrems)** [#5](https://github.com/testiumjs/testium-example-app/pull/5)
  - [`ad7d82e`](https://github.com/testiumjs/testium-example-app/commit/ad7d82ea92d4e155b0c7fe8222c9774a4223de5f) **chore:** Ship public and server to npm


### 2.0.1

* Apply latest nlm generator - **[@i-tier-bot](https://github.com/i-tier-bot)** [#4](https://github.com/testiumjs/testium-example-app/pull/4)
  - [`52fc9ac`](https://github.com/testiumjs/testium-example-app/commit/52fc9acf4e88eb598e3370d27f2db9515459f3f4) **chore:** Apply latest nlm generator
  - [`a3c7aa7`](https://github.com/testiumjs/testium-example-app/commit/a3c7aa7fbe889e099055232fb82eed51ea75ed2c) **chore:** Apply snowflake changes


test/testium-example-app.test.js
'use strict';
var assert = require('assertive');

var testiumExampleApp = require('../');

describe('testium-example-app', function () {
  it('is empty', function () {
    assert.deepEqual({}, testiumExampleApp);
  });
});


server.js
#!/usr/bin/env node

'use strict';

const http = require('http');
const path = require('path');
const fs = require('fs');
const console = require('console');

function echo(request, response) {
  const data = {
    ip: request.connection.remoteAddress,
    method: request.method,
    url: request.url,
    body: '',
    headers: request.headers,
  };

  request.on('data', buffer => {
    data.body += buffer.toString();
  });

  request.on('end', () => {
    const body = JSON.stringify(data, null, 2);
    response.setHeader('Content-Type', 'application/json');
    response.end(body);
  });
}

function error(response) {
  response.statusCode = 500;
  response.end('500 SERVER ERROR');
}

function crash(response) {
  response.statusCode = 500;
  response.socket.destroy();
}

function serveFromDisk(pathname, response) {
  let safePath = pathname.replace(/\.\./g, '').replace(/^\/+/, '');
  if (safePath === '') safePath = 'index.html';
  const filePath = path.resolve(__dirname, 'public', safePath);
  if (fs.existsSync(filePath) && fs.statSync(filePath).isFile()) {
    let contentType = 'text/html';
    if (filePath.endsWith('.js')) {
      contentType = 'application/javascript';
    }
    response.setHeader('Content-Type', contentType);
    fs.createReadStream(filePath).pipe(response);
  } else {
    response.statusCode = 404;
    response.end('File Not Found');
  }
}

function createServer() {
  return http.createServer((request, response) => {
    const { pathname } = new URL(request.url, `http://${request.headers.host}`);
    switch (pathname) {
      case '/echo':
        return echo(request, response);

      case '/error':
        return error(response);

      case '/crash':
        return crash(response);

      case '/blackhole':
        return null;

      default:
        return serveFromDisk(pathname, response);
    }
  });
}

const testApp = (module.exports = {
  listen: function listen(port, callback) {
    this.server = createServer();
    this.server.listen(port, callback);
  },

  kill: function kill(callback) {
    this.server.close(callback);
    this.server = null;
  },
});

if (module === require.main) {
  if (process.env.never_listen) {
    console.log('Refusing to listen');
    setTimeout(() => {}, 100000);
  } else {
    testApp.listen(process.env.PORT || 4003, function onListen() {
      console.log('Listening on port %j', this.address().port);
    });
  }
}
